### PR TITLE
NEW: Advanced jobs admin UI.

### PR DESCRIPTION
**Pull request by mfendeksilverstripe**
# Advanced jobs admin UI

Optional advanced UI for jobs admin. Needs to be enabled via configuration API.

## Problem
Finding specific jobs in the Jobs admin is a chore, developers often find themselves scrolling through multiple pages. Finding the right jobs is sometimes impossible as the filter criteria input is quite limited. Using Text fields for input is often tedious.

## Solution
Improved UI. More filters, better input fields, more useful information in the columns.

job-admin

## Notes

* split from [Feature batch](https://github.com/symbiote/silverstripe-queuedjobs/pull/290) as `Feature 1 - Improved Jobs admin UI`

## Related issues
https://github.com/symbiote/silverstripe-queuedjobs/issues/308

### CODE REVIEW

**Comment by sachajudd** [comment]
@silverstripeux

File: src/DataObjects/QueuedJobDescriptor.php
Comment (emteknetnz) [inline src/DataObjects/QueuedJobDescriptor.php]: The use of the word 'summary' is a little weird, perhaps rename it to `getTrimmedImplementation()` or similar?.  I'd also add example to show that its trimming off the front of the namespace

File: README.md
Comment (emteknetnz) [inline README.md]:
```suggestion
## Enabling advanced admin UI
```

File: src/Controllers/QueuedJobsAdmin.php
Comment (emteknetnz) [inline src/Controllers/QueuedJobsAdmin.php]:
```suggestion
     * This requires the optional module
```

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Look great, you'll need to rebase/retarget this to `4` since it contains new API

File: README.md
Comment (sachajudd) [reply README.md]: How about 'Enabling extended admin UI' instead of 'Advanced'? Even though this is the readme, we try to avoid advanced wording coupled in UI contexts where possible.

**Review by Cheddam — CHANGES_REQUESTED** [review]
Can you also comb through and add translation calls for all the new language strings?

**Comment by mfendeksilverstripe** [comment]
@emteknetnz @Cheddam Rebased on `4`, pushed up new changes, please review.

**Comment by emteknetnz** [comment]
@mfendeksilverstripe I tried testing this locally though it didn't seem to activate the filterable headers.  I added the yml and installed the suggested module and ran /dev/build?flush

It's possible that it's related to merge conflicts after merging https://github.com/symbiote/silverstripe-queuedjobs/pull/314 - could you rebase this and see if it fixes things up?

**Comment by mfendeksilverstripe** [comment]
@emteknetnz I rebased the branch. Also, I tested it on my local and it seems to be working fine. Have you enabled the UI on your local? It's disabled by default so you need to add following config to your YAML:

```
Symbiote\QueuedJobs\Controllers\QueuedJobsAdmin:
  advanced_admin_ui: true
```

**Comment by emteknetnz** [comment]
@mfendeksilverstripe yeah I've required `silverstripe-terraformers/gridfield-rich-filter-header` though it does not appear to make any difference on at the top-level view `/admin/queuedjobs/`

I have included the following config and run /dev/build?flush

```yaml
Symbiote\QueuedJobs\Controllers\QueuedJobsAdmin:
  advanced_admin_ui: true
```

Also looks like you'll need to do another rebase (sorry) after I merged another one of your queuedjob prs

**Comment by mfendeksilverstripe** [comment]
@emteknetnz I have rebased the branch and tested this feature on a vanilla install. It seems to be working fine. Here is my project config

Screen Shot 2020-09-25 at 8 48 28 AM

Note that I replaced the queued jobs module folder with this branch manually.

I recommend checking the version of the rich filter header module and also the following setting:

Screen Shot 2020-09-25 at 8 48 43 AM

This setting should be automatically set by the rich filter header module.

**Comment by madmatt** [comment]
^ @emteknetnz Did you happen to get anywhere with this? Looks like a great addition and I'd be keen to see it merged in (noting there's a new conflict already heh)

**Comment by mfendeksilverstripe** [comment]
@madmatt I've rebased the branch so it should be easy to merge in now.

**Comment by emteknetnz** [comment]
@mfendeksilverstripe I must be missing something?  I've included this PR, `silverstripe-terraformers/gridfield-rich-filter-header:^2` and yml config in my project and run dev/build flush - the queued jobs ui still looks exactly the same?  I've run it through xdebug and it's going all the way through `enableAdvancedAdminUi()`

Here's mysite.yml + composer.json

```
---
Name: myproject
---
SilverStripe\Core\Manifest\ModuleManifest:
  project: app

# UTF8MB4 has limited support in older MySQL versions.
# Remove this configuration if you experience issues.
---
Name: myproject-database
---
SilverStripe\ORM\Connect\MySQLDatabase:
  connection_charset: utf8mb4
  connection_collation: utf8mb4_unicode_ci
  charset: utf8mb4
  collation: utf8mb4_unicode_ci


Symbiote\QueuedJobs\Controllers\QueuedJobsAdmin:
  advanced_admin_ui: true
```

```
{
    "name": "silverstripe/installer",
    "type": "silverstripe-recipe",
    "description": "The SilverStripe Framework Installer",
    "require": {
        "php": "^7.1 || ^8",
        "silverstripe/recipe-plugin": "^1.2",
        "silverstripe/recipe-cms": "4.x-dev",
        "silverstripe-themes/simple": "~3.2.0",
        "silverstripe/login-forms": "4.x-dev",
        "symbiote/silverstripe-queuedjobs": "dev-feature/jobs-admin-ui as 4.x-dev",
        "silverstripe/recipe-content-blocks": "2.x-dev",
        "silverstripe/graphql": "^3@dev",
        "symfony/yaml": "^3",
        "silverstripe-terraformers/gridfield-rich-filter-header": "^2"
    },
    "require-dev": {
        "sminnee/phpunit": "^5.7",
        "sminnee/phpunit-mock-objects": "^3.4.5",
        "silverstripe/recipe-testing": "^1"
    },
    "extra": {
        "project-files": [
            "app/_config/*",
            ".env.example"
        ],
        "public-files": [
            "assets/*",
            "favicon.ico"
        ],
        "resources-dir": "_resources",
        "project-files-installed": [
            "app/.htaccess",
            "app/_config.php",
            "app/_config/contentblocks.yml",
            "app/_config/mimevalidator.yml",
            "app/_config/mysite.yml",
            "app/src/BlocksPage.php",
            "app/src/Page.php",
            "app/src/PageController.php",
            "behat.yml",
            "phpcs.xml.dist",
            "phpunit.xml.dist"
        ],
        "public-files-installed": [
            ".htaccess",
            "index.php",
            "web.config"
        ]
    },
    "config": {
        "process-timeout": 600,
        "platform": {
            "php": "7.2"
        }
    },
    "prefer-stable": true,
    "minimum-stability": "dev",
    "repositories": {
        "x1": {
            "type": "vcs",
            "url": "git@github.com:silverstripe-terraformers/silverstripe-queuedjobs.git"
        }
    }
}

```

**Comment by mfendeksilverstripe** [comment]
Hey @emteknetnz here is a step by step guide which I used to test this feature on a vanilla install. Hopefully, this helps :)

1 - `composer create-project silverstripe/installer vanilla ^4`
2 - add `.env` and `vagrant` files
3 - `vagrant up`
4 - `composer install`
5 - `dev/build` with flush
6 - add fork to the composer file
```
"repositories": {
    "silverstripe/queuedjobs": {
        "type": "git",
        "url": "https://github.com/silverstripe-terraformers/silverstripe-queuedjobs.git"
    }
},
```
7 - `composer require symbiote/silverstripe-queuedjobs`
8 - change the queued jobs module in your composer JSON to point to the fork:

`"symbiote/silverstripe-queuedjobs": "dev-feature/jobs-admin-ui as 4.x-dev",`

9 - `composer update symbiote/silverstripe-queuedjobs`
10 - `composer require silverstripe-terraformers/gridfield-rich-filter-header`
11 - Enable the advanced UI via static configuration

```
Symbiote\QueuedJobs\Controllers\QueuedJobsAdmin:
  advanced_admin_ui: true
```
12 - `dev/build` with flush

This is what I see when I navigate to the Queued jobs admin:

Screen Shot 2021-10-18 at 2 02 33 PM

cc: @madmatt

**Comment by emteknetnz** [comment]
Nope I'm still not seeing it.  The steps you've given are basically the same as the sample composer.json and mysite.yml I provided above.  I've also tried ^2 and dev-master of silverstripe-terraformers/gridfield-rich-filter-header though it made no difference

Is there some extra config missing in either your .env file of vagrantfile?

**Comment by mfendeksilverstripe** [comment]
@emteknetnz  I don't think my env file is related to this but here it is just for reference:

```
SS_ENVIRONMENT_TYPE="dev"

SS_DATABASE_CLASS="MySQLDatabase"
SS_DEFAULT_ADMIN_USERNAME="admin"
SS_DEFAULT_ADMIN_PASSWORD="password"
SS_DATABASE_SERVER="localhost"
SS_DATABASE_USERNAME="root"
SS_DATABASE_PASSWORD="root"
SS_DATABASE_NAME="SS_vanilla"

SS_DATABASE_TIMEZONE="+00:00"
SS_BASE_URL="http://vanilla.local/"
SS_SEND_ALL_EMAILS_TO=""
```

I don't have anything out of ordinary in the vagrant file either. I'll try to add a unit test which would validate that the UI is loading properly.

**Comment by mfendeksilverstripe** [comment]
Hey @emteknetnz I've added unit tests to showcase that the new UI is working as expected. I recommend running the test suite on your local setup and see if it passes.

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Apologies, looking again at this with fresh eyes I'm really not sure that we should be essentially baking an third party optional module into queuedjobs.  I feel like things should be implemented in a bit more of an "extendable" fashion.

Most core maintainers won't have merge access on the terraformers account.  It seems like if we're going to integrate this module to this degree we'd probably need everything on the Silverstripe account.  However that goes against the idea that functionality like this would be better maintained outside of the core team

It also seems like you could simply use the existing `$this->extend('updateEditForm', $form);` hook within the terraformers module to to do everything you need to ... does this seem correct?

I'm happy with linking to the module within docs.  Not so sure about `suggest` in composer.json, does that still work in composer 2?

**Comment by mfendeksilverstripe** [comment]
Thanks for clarification, I'll close this PR then.

**Pull request closed without merging**
